### ISSUE TITLE
Can and should we overload assignment `:=`?

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#4378]

**Comment by JonathanStarup** [comment]
It could use an Assignable typeclass. Making it open to general regioned data structures

**Comment by magnus-madsen** [comment]
`Index` and `IndexMut` https://doc.rust-lang.org/std/ops/trait.IndexMut.html

Perhaps:

```
class Index[t: Type -> Region -> Type] {
    pub def index(i: Int32, m: t[a, r]): a \ r
}

class IndexMut[t: Type -> Region -> Type] {
    pub def indexMut(x: a, i: Int32, m: t[a, r]): Unit \ r
}
```

**Comment by mlutze** [comment]
What happens when we have a mutable list of indexes?

```scala
list[42] := "hello"
list[42] := ref "hello"
```

Seems like it might get confusing.

**Comment by magnus-madsen** [comment]
Why? Its always assignment to the LHS.

**Comment by magnus-madsen** [comment]
Does it get more confusing than in other languages?

**Comment by mlutze** [comment]
> Why? Its always assignment to the LHS.

Right, but are you assigning to ref at index `i` or replacing the ref at index `i`?

> Does it get more confusing than in other languages?

I'm not sure what other languages have this problem, since often languages with mutability use mutable fields rather than references. OCaml has a separate syntax for assigning to references vs assigning to arrays. (`:=` vs `<-`)
